src/utils.rs

```diff
@@ -825,85 +823,114 @@ fn get_meta_info(
+
+fn parse_punctuated_nested_meta(
+    info: &mut MetaInfo,
+    meta: &Punctuated<NestedMeta, Token![,]>,
+    allowed_attr_params: &[&str],
+    value: bool,
+) -> Result<()> {
+    for meta in meta.iter() {
+        let meta = match meta {
+            NestedMeta::Meta(meta) => meta,
+            _ => {
+                return Err(Error::new(meta.span(), "Attribute format not supported2"))
             }
-            "ref" => {
-                info.ref_ = Some(true);
+        };
+
+        match meta {
+            Meta::List(list) if list.path.is_ident("not") => {
+                if value {
+                    parse_punctuated_nested_meta(
+                        info,
+                        &list.nested,
+                        allowed_attr_params,
+                        false,
+                    )?;
+                } else {
+                    return Err(Error::new(
+                        meta.span(),
+                        "Attribute format not supported3",
+                    ));
+                }
             }
-            "ref_mut" => {
-                info.ref_mut = Some(true);
+
+            Meta::Path(path) => {
+                if !allowed_attr_params.iter().any(|param| path.is_ident(param)) {
+                    return Err(Error::new(
+                        meta.span(),
+                        format!(
+                            "Empty attribute is not allowed, add one of the following parameters: {}",
+                            allowed_attr_params.join(", "),
+                        ),
+                    ));
+                }
+
+                // not(ignore) does not make much sense
+                if path.is_ident("ignore") && value {
+                    info.enabled = Some(false);
+                } else if path.is_ident("forward") {
+                    info.forward = Some(value);
+                } else if path.is_ident("owned") {
+                    info.owned = Some(value);
+                } else if path.is_ident("ref") {
+                    info.ref_ = Some(value);
+                } else if path.is_ident("ref_mut") {
+                    info.ref_mut = Some(value);
+                } else if path.is_ident("source") {
+                    info.source = Some(value);
+                } else {
+                    return Err(Error::new(
+                        meta.span(),
+                        "Attribute format not supported4",
+                    ));
+                };
             }
+
             _ => {
-                return Err(Error::new(meta.span(), "Attribute format not supported7"));
+                return Err(Error::new(meta.span(), "Attribute format not supported5"))
             }
         }
     }
-    Ok(info)
+
+    Ok(())
 }
 
-#[derive(Copy, Clone, Debug)]
+#[derive(Copy, Clone, Debug, Default)]
 pub struct FullMetaInfo {
     pub enabled: bool,
     pub forward: bool,
     pub owned: bool,
     pub ref_: bool,
     pub ref_mut: bool,
+    pub info: MetaInfo,
```

Comment:
@JelteF I'd like to have an ability to write
```rust
#[derive(Error)]
#[error(not(source))]
struct WithoutSource(i32)
```
 _in addition to_
```rust
#[derive(Error)]
struct WithoutSource(#[error(not(source))] i32)
```
but not _instead of_. We have a lot of multi-fielded tuple variants and putting attribute on a field rather than on a variant is vital for our case.

------------------

> To make this last one convenient with backtrace we might need an auto modifier in addition to the not modifier to set it the field to None again for a specific variant.

Didn't get it. Would you be so kind to provide an example with `auto`? For me it seems quite redudant atm.

Reply:
Yes, this would be in addition to putting it on an attribute. The result of putting it on the the variant, would be the same as putting it on all of the fields.

An example of the `auto` modifier would be:
```rust
#[derive(Error)]
#[error(not(source))]
enum CompoundError {
    #[error(auto(source))]
    WithSource {
        source: Simple,
        backtrace: Backtrace,
    },
    #[error(auto(source))]
    Tuple(WithExplicitSource, Backtrace),
    WithoutSource(Tuple),
    WithoutSource1(Tuple),
    WithoutSource2(Tuple),
}
```

Simply using `#[error(source)]` on the variant is not possible here, since it would set `source` true for all fields (also the `Backtrace` one). The `auto` would set the source field back to `None` from `false`. However, I'm not sure if there isn't a nicer solution, so let's decide on this later.

Reply:
@JelteF we cannot use `Backtrace` since it doesn't impl `Error` itself, so this will result always in compilation error. It's better to automatically omit `Backtrace`s when resolving `source`, but in either way it's a theme of a separate PR as this one doesn't introduce any `Backtrace`s.

------------------------------------

> The `auto` would set the source field back to `None` from `false`.

I'm sorry, but I still haven't got it. At the moment I see the following picture:
```rust
#[derive(Error)]
#[error(not(source))] // instructs to return `None` for all variants by default
enum CompoundError {
    WithSource {
        #[error(source)] // instructs to return this field as source for this variant, overrides any defaults
        source: Simple,
        extra_info: bool,
    },
    #[error(source)] // instructs to detect source for this variant via default logic
    Tuple(WithExplicitSource),
    #[error(source)] // instructs to detect source for this variant via default logic
    WithAutoSource {
        source: Simple,
        backtrace: Backtrace,
    },
    #[error(source)] // instructs to detect source for this variant via default logic, 
                     // but that is impossible in fact due to ambiguity,
                     // should be a macro expansion error requiring to place attribute on a field
    Tuple2(WithExplicitSource, AnotherSource),
    WithoutSource(Tuple),
    WithoutSource1(Tuple),
    WithoutSource2(Tuple),
}
```
I cannot see any sense in assuming `source` true for all fields. Even if the error really holds 2 sources inside, it still can return the only one from a `source()` method.  
Yeah, of course we can assume the situation like `With2Sources(Option<Source1Error>, Option<Source2Error>)`, but that smells like the anti-pattern and _should be_ refactored into two separate variants. But should we consider anti-patterns?

Reply:
Hmm... looking at the code pieces above I'm unsure now whether it worths to put `source` attribute at variants at all. It feels quite confusing for readability and complicates understanding what's going on. I'd prefer to have them on fields only and, optionally, on the whole type for changing defaults for all variants (for better DRY in some cases), but not on per-variant basis.

Reply:
Alright, you're making some good points. Let's keep it as is for now. And we can refine this in a future issue+PR without breaking the API because right now placing attributes anywhere but on fields results in an error.